Implement pipe to daemon

BryanCrotaz opened the pull request:
Implement #50 
Use as normal, only change is in bootstrap call:

```
// old behaviour, requires sudo to run
Pi.Init<BootstrapPiGpio<InProcessCommsStrategy>>();

// communicate with daemon over pipe
Pi.Init<BootstrapPiGpio<PipeCommsStrategy>>();

// communicate with daemon over socket - TODO
Pi.Init<BootstrapPiGpio<SocketCommsStrategy>>();
```

Commit:
Implement to bring in line with the WiringPi implementation

Commit:
implement pin lookup

Commit:
implement pin lookups

Commit:
fix typo

Commit:
move in-process startup to a strategy class

Commit:
Add strategies for communicating with pigpio

Commit:
implement basic gpio functions on pipe communication to daemon

BryanCrotaz commented:
Work in progress, help appreciated!

Commit:
clean up

Commit:
Extract pipe wrappers ready for notification pipes

Commit:
Add ISR hooks via pipe notifications

Commit:
fix build

Commit:
Add threads service and refactor pipe access

Commit:
refactor utilities service

Commit:
Refactor flags extension methods out of native methods

Commit:
compiler hints

Commit:
Fix null ref warning

Commit:
implement utility service pipe

Commit:
Fix compiler warning

Commit:
Add services for PWM and implement pipe version

Commit:
Clean termination

Commit:
implement Terminate placeholder on socket

Commit:
Use new PWM service

Commit:
Implement serial service

Commit:
complete serial service implementation

Commit:
implement waves service

Commit:
Implement I2C service

Commit:
start implementation of I2C

BryanCrotaz commented:
Most important pieces are now implemented, with `NotImplementedException` on those that aren't yet. In Process functionality should not be altered.

BryanCrotaz commented:
@k3z0 what do you think?

BryanCrotaz commented:
@geoperez is this something you can accept as a PR?

geoperez commented:
Hello @BryanCrotaz , first thanks for your contribution. It's a really big PR, but @k3z0 is taking care of the code review :)

BryanCrotaz commented:
The concept is simple:

Abstract the public interface and load an implementation through the dependency system
Implementations are InProcess (the existing code), Pipe (mostly implemented here), Socket (not yet inplemented)

Advantages - Pipe & socket don't need su rights, and socket can be run from a different machine

`IPiGpioCommsStrategy` sets up the services for a particular implementation

Commit:
add set to gpio pwm service so you can control pwm
Also clean up BoardException to always return positive result type (negative is error which maps to throw)

Commit:
fix pwm and serial over pipe

Commit:
run code cleanup

Commit:
add PWM flag to gpio 10

k3z0 commented on Unosquare.PiGpio.Workbench/Runners/DhtSensor.cs:
```diff
@@ -78,7 +78,7 @@ protected override void DoBackgroundWork(CancellationToken ct)
                     bitArray.CopyTo(bytes, 0);
 
-                    $"Bytes Received: {BitConverter.ToString(bytes).Replace("-", " ")}".Info(Name);
+                    $"Bytes Received: {BitConverter.ToString(bytes).Replace("-", " ", StringComparison.InvariantCulture)}".Info(Name);
```
```suggestion
                    $"Bytes Received: {BitConverter.ToString(bytes).Replace("-", " ", StringComparison.Ordinal)}".Info(Name);
```

k3z0 commented on Unosquare.PiGpio.Workbench/Runners/LedBlinking.cs:
```diff
@@ -13,22 +13,22 @@ public LedBlinking(bool isEnabled)
         {
-            _pin.Value = false;
-            Board.Timing.Sleep(200);
+            _pin.Value = !_pin.Value;
```
This should be explicitly false, since you don't know the current status of the pin when the process ends

k3z0 commented on Unosquare.PiGpio.Workbench/Runners/Servo.cs:
```diff
@@ -23,7 +23,7 @@ protected override void OnSetup()
         {
             var pulseWidth = InitialPulseWidth;
-            var pulseDelta = 10;
+            int pulseDelta = 10;
```
```suggestion
            var pulseDelta = 10;
```

k3z0 commented on Unosquare.PiGpio/NativeTypes/GpioPulse.cs:
```diff
@@ -4,8 +4,7 @@
     /// <summary>
-    /// A pulse representing microseconds in the high position,
-    /// microseconds in the low position, and a delay measure.
+    /// A waveform entry consisting of GPIO bits to turn on, GPIO bits to turn off, and a delay before the next waveform entry.
```
```suggestion
    /// A waveform entry consisting of GPIO bits to turn on, GPIO bits to turn off, and a delay before the next pulse.
```

BryanCrotaz replied:
pulse implies a pair of edges. A wave form entry is one edge. I'd prefer to leave this one as is

k3z0 commented:
Hello @BryanCrotaz , 

Thanks for your contribution. 

This is big PR, I need to review your code and run some test using the Raspberry, so it will take me some time to finish all the tests.

BryanCrotaz commented:
Of course.

I'm using _some_ of the functionality via pipe (levels, notifications and PWM) and it's working really nicely. I've tested the levels and pwm via in-process which shouldn't have changed - it's just an extra level of indirection.

Commit:
fix code review issues

BryanCrotaz commented:
I've left lots as NotImplemented, for others to contribute. I've got a socket implementation in another project that I'll copy over when I get a moment (as a separate PR once this is accepted).

BryanCrotaz commented:
> Hello @BryanCrotaz ,
> 
> Thanks for your contribution.
> 
> This is big PR, I need to review your code and run some test using the Raspberry, so it will take me some time to finish all the tests.

Is the restructure making sense?

BryanCrotaz commented:
I'm now using this in production. It's working well.

geoperez commented:
@k3z0 what do you think?

BryanCrotaz commented:
Any news?

BryanCrotaz commented:
Nudge

BryanCrotaz commented:
Nudge

geoperez commented:
cc @k3z0

BryanCrotaz commented:
I've been using this in production now for 6 months. I'd prefer not to be running a fork - what more do you need from me to be able to verify this?

geoperez commented:
We are looking for a new maintainer, apply at https://adoptoposs.org/p/d3470190-942b-44ac-84fc-90259cbdee43

Commit:
fix race condition in GPIO level change notifications
